sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1430,6 +1430,11 @@ class Dataset[T] private[sql](
    */
   @scala.annotation.varargs
   def select(cols: Column*): DataFrame = withPlan {
```

Comment:
is it possible to support mixed typed and untyped columns?

Reply:
For typed column, there is encoder which specifies how we can encode its output. We have no such info for untyped columns.

Reply:
Hm, but previously `count` case was possible to use it as untyped column.

```scala
scala> spark.range(1).select(count("id"), count("id"), count("id"), count("id"), count("id"), count("id")).show()
+---------+---------+---------+---------+---------+---------+
|count(id)|count(id)|count(id)|count(id)|count(id)|count(id)|
+---------+---------+---------+---------+---------+---------+
|        1|        1|        1|        1|        1|        1|
+---------+---------+---------+---------+---------+---------+


scala> count("id")
res0: org.apache.spark.sql.TypedColumn[Any,Long] = count(id)
```

Reply:
Ah, but problem is more specific to `Aggregator` because users can set the type explicitly and later `encoder` is required, specifically via `TypedColumn.withInputType` -> `TypedAggregateExpression.withInputInfo`

Reply:
What about a fix like this?

```diff
NOTICE-binary                       common                              examples                            metastore_db                        sbin                                streaming
diff --git a/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala b/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala
index 16f1cac3e0f..91b6e13b100 100644
--- a/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala
+++ b/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala
@@ -1430,12 +1430,11 @@ class Dataset[T] private[sql](
    */
   @scala.annotation.varargs
   def select(cols: Column*): DataFrame = withPlan {
-    cols.find(_.isInstanceOf[TypedColumn[_, _]]).foreach { typedCol =>
-      throw new AnalysisException(s"$typedCol is a typed column that " +
-        "cannot be passed in untyped `select` API. If you are going to select " +
-        "multiple typed columns, you can use `Dataset.selectUntyped` API.")
+    val newCols = cols.map {
+      case tc: TypedColumn[_, _] => tc.withInputType(exprEnc, logicalPlan.output)
+      case c => c
     }
-    Project(cols.map(_.named), logicalPlan)
+    Project(newCols.map(_.named), logicalPlan)
```

Reply:
It cannot work because typed column's output is a domain object instead of a column. That is why untyped `select` returns DataFrame and typed `select` returns Dataset[U1], Dataset[(U1, U2)], Dataset[(U1, U2, U3)]...

`count` is possible before because it is actually untyped column but wrongly being wrapped in a typed column.

Reply:
Hm.. I see. So to support typed columns properly, we should either expose `selectUntyped` or make all cases for `def select[U1, ...](`  ..

Reply:
Yea, I think so if I do not miss anything.